### PR TITLE
Kamu UI 307 implementation of ingest via file upload from UI

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**
Result:

Image

Image

https://github.com/kamu-data/kamu-web-ui/assets/51016717/73cfb7b0-19fa-4e72-86b1-6f4f473da36d

Image

Image

### CODE REVIEW

File: src/app/common/app.types.ts [hunk 1]
```diff
@@ -1,3 +1,22 @@
...
+    fields: [string, string][];
+}
+
+export interface UploadPerareData {
+    uploadPrepareResponse: UploadPrepareResponse;
+    bodyObject: FormData | File;
+    uploadHeaders: HttpHeaders;
+}
+
+export type UploadAvailableMethod = "POST" | "PUT";
```
Comment (zaychenko-sergei) [inline src/app/common/app.types.ts:22]: It deserves a separate file

File: src/app/dataset-view/additional-components/overview-component/overview.component.scss [hunk 2]
```diff
@@ -5,3 +5,56 @@
...
+    // .file-uploader {
+    //     opacity: 0;
+    //     padding: 4.5rem;
+    //     width: inherit;
+    //     height: 100%;
+
+    //     &:hover {
+    //         cursor: pointer;
+    //     }
+    // }
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/overview-component/overview.component.scss:38]: Garbage

File: src/app/services/file-upload.service.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import { NavigationService } from "./navigation.service";
+
+@Injectable({
+    providedIn: "root",
+})
+export class FileUploadService {
+    private authHeaders: MaybeUndefined<{
+        [header: string]: string | string[];
+    }>;
+    private uploadToken: string;
```
Comment (zaychenko-sergei) [inline src/app/services/file-upload.service.ts:22]: The headers are defined in constructor, but not the token

File: src/app/services/file-upload.service.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        const cache = this.injector.get(APOLLO_OPTIONS).cache;
+        if (cache) {
+            const datasetKeyFragment = DatasetApi.generateDatasetKeyFragment(
+                cache.identify(DatasetApi.generateAccountKeyFragment(datasetBasics.owner.id)),
+                datasetBasics.id,
+            );
+            cache.evict({
+                id: cache.identify(datasetKeyFragment),
+                fieldName: "data",
+            });
```
Comment (zaychenko-sergei) [inline src/app/services/file-upload.service.ts:90]: I think "metadata" section should be reset too.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
Things to add:
1) Error handling. Upload prepare, actual upload and ingest - all these phases may fail. Currently erors are not even displayed in the Toastr.


2) Displaying busy indicator, perhaps modal, but decide on your own - we allow uploading large files (50mb) and ingestion of those takes a while too.


3) Control upload file limit. Let's take the value from runtime configuration too, have 50mb by default. `(done)`

4) When `SetPollingSource` is defined, we should hide the block with drag & drop. The same for derivative datasets. `(done)`

5) Not enough tests

File: src/app/common/app.types.ts (hunk 1 above)
Comment (dmitriy-borzenko) [reply src/app/common/app.types.ts:22]: Moved to a separate file

File: src/app/dataset-view/additional-components/overview-component/overview.component.scss (hunk 2 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/overview-component/overview.component.scss:38]: Removed

File: src/app/services/file-upload.service.ts (hunk 3 above)
Comment (dmitriy-borzenko) [reply src/app/services/file-upload.service.ts:22]: Removed 2 variables from service

File: src/app/services/file-upload.service.ts (hunk 4 above)
Comment (dmitriy-borzenko) [reply src/app/services/file-upload.service.ts:90]: Replace on `metadata`

**Review by zaychenko-sergei — APPROVED** [review]

**Pull request merged**